### PR TITLE
Add SafeExecutor and SafeScheduledExecutor.

### PR DESCRIPTION
**Pull request by bgrozev**

### CODE REVIEW

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+class ExecutorShutdownTimeoutException : Exception("Timed out trying to shutdown executor service")
+
+class SafeExecutor(
+    name: String,
+    delegate: ExecutorService
+) : TaskServiceWrapper<ExecutorService>(name, delegate) {
+
+    fun execute(task: Runnable) = delegate.execute(wrapTask(task))
```
Comment (bbaldino) [inline jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: It looks like you mimic the `Executor` interface here, but also have `ExecutorService`-like methods (the submit-style methods).  Why not just mimic the `ExecutorService` API (which is what we use everywhere) for both types and have the `submit` calls and `unsafeSubmit` variants?

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    fun <T> unsafeSubmit(task: Runnable, result: T): Future<T>? = delegate.submit(task, result)
+    fun unsafeSubmit(task: Runnable): Future<*>? = delegate.submit(task)
+}
+
+class SafeScheduledExecutor(
+    name: String,
+    delegate: ScheduledExecutorService
+) : TaskServiceWrapper<ScheduledExecutorService>(name, delegate) {
+
+    fun schedule(command: Runnable, delay: Long, unit: TimeUnit) {
```
Comment (bbaldino) [inline jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:97]: Do we need a `KFunction0<Unit>` version of this?

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        unsafeSchedule(wrapTask(command), delay, unit)
+    }
+
+    fun unsafeSchedule(command: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture<*> =
+        delegate.schedule(command, delay, unit)
+
+    fun unsafeSchedule(command: KFunction0<Unit>, delay: Duration): ScheduledFuture<*> =
+        unsafeSchedule(Runnable { command() }, delay.toMillis(), TimeUnit.MILLISECONDS)
+
+    fun unsafeScheduleAtFixedRate(
```
Comment (bbaldino) [inline jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:107]: What about safe versions of `scheduleAtFixedRate` and `scheduleWithFixedDelay` ?

File: jitsi-utils/src/main/java/org/jitsi/utils/queue/AsyncQueueHandler.java [hunk 4]
```diff
@@ -45,16 +45,23 @@
      * The default <tt>ExecutorService</tt> to run <tt>AsyncQueueHandler</tt>
      * when there is no executor injected in constructor.
      */
-    private final static ExecutorService sharedExecutor
-        = Executors.newCachedThreadPool(
-            new CustomizableThreadFactory(
-                AsyncQueueHandler.class.getName() + "-executor-", true));
+    private final static Executor sharedExecutor = new Executor() {
```
Comment (bbaldino) [inline jitsi-utils/src/main/java/org/jitsi/utils/queue/AsyncQueueHandler.java:48]: So this is not using the safe variant?

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 1 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: `ExecutorService` extends `Executor` (so it inherits `void execute`), which is why I named the "safe" method `execute`.

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 2 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:97]: Sure

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 3 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:107]: I can add all of these, thought I don't know how useful they are.

File: jitsi-utils/src/main/java/org/jitsi/utils/queue/AsyncQueueHandler.java (hunk 4 above)
Comment (bgrozev) [reply jitsi-utils/src/main/java/org/jitsi/utils/queue/AsyncQueueHandler.java:48]: No, it uses the `Future`. It does it's own exception handling, at [least for the `handleItem` method](https://github.com/jitsi/jitsi-utils/blob/master/jitsi-utils/src/main/java/org/jitsi/utils/queue/AsyncQueueHandler.java#L139)

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 1 above)
Comment (bbaldino) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: Sorry, I don't think I followed this.  Why expose `execute` at all, given nothing we use uses `Executor` directly, but instead `ExecutorService`/`ScheduledExecutorService`?

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 3 above)
Comment (bbaldino) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:107]: It looks like we use `scheduleAtFixedRate` in `Conference.java` in the bridge?

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 1 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: To mimic the `ExecutorService` API -- the method which runs a task and doesn't return anything is named `execute`, while `submit` methods return a future.

Comment (bbaldino) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: Ok, from what I saw we didn't use `execute` anywhere in the code, so think I was confused.  Are you thinking that we should change the places which don't use the future over to `execute` (from `submit`)? That's probably a good idea.

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 3 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:107]: Yes, that's now `unsafeScheduleAtFixedRate`

File: jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt (hunk 1 above)
Comment (bgrozev) [reply jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt:84]: Yes, that's what I did in the PRs in jmt/jvb

**Pull request closed without merging**